plugins/kafka/kafka.hpp

```diff
@@ -0,0 +1,447 @@
+
+#ifdef ECL_KAFKA_EXPORTS
+extern "C"
+{
+    ECL_KAFKA_API bool getECLPluginDefinition(ECLPluginDefinitionBlock *pb);
+}
+#endif
+
+extern "C++"
+{
+    namespace KafkaPlugin
+    {
+        class KafkaObj;
+        class Poller;
+        class Publisher;
+        class Consumer;
+        class KafkaStreamedDataset;
+
+        /** @class KafkaObj
+         *
+         *  Parent class for both Publisher and Consumer classes.  Provides
+         *  easy way for a Poller object to access either for callbacks, etc.
+         */
+        class KafkaObj
+        {
+            public:
+
+                /**
+                 * Returns a pointer to the librdkafka object that can be either
+                 * a producer or consumer.
+                 */
+                virtual RdKafka::Handle* handle() = 0;
+        };
+
+        //----------------------------------------------------------------------
+
+        /** @class Poller
+         *
+         *  Background execution of librdkafka's poll() function, which is
+         *  required in order to batch I/O.  One Poller will be created
+         *  for each Publisher and Consumer object actively used
+         */
+        class Poller : public Thread
+        {
+            public:
+
+                /**
+                 * Constructor
+                 *
+                 * @param   _parentPtr      Pointer to Publisher or Consumer object
+                 *                          that created this object
+                 * @param   _pollTimeout    The number of milliseconds to wait
+                 *                          for events within librdkafka
+                 */
+                Poller(KafkaObj* _parentPtr, __int32 _pollTimeout);
+
+                virtual ~Poller();
+
+                /**
+                 * Starts execution of the thread main event loop
+                 */
+                virtual void start();
+
+                /**
+                 * Stops execution of the thread main event loop.  Note that we
+                 * wait until the main event loop has actually stopped before
+                 * returning.
+                 */
+                void stop();
+
+                /**
+                 * Entry point to the thread main event loop.  Exiting this
+                 * method means that the thread should stop.
+                 */
+                virtual int run();
+
+            private:
+
+                bool        shouldRun;      //!< If true, we should execute our thread's main event loop
+                KafkaObj*   parentPtr;      //!< Pointer to object that started this threaded execution
+                __int32     pollTimeout;    //!< The amount of time (in ms) we give to librdkafka's poll() function
+        };
+
+        //----------------------------------------------------------------------
+
+        class Publisher : public virtual KafkaObj, public virtual RdKafka::EventCb, public virtual RdKafka::DeliveryReportCb
```

Comment:
Do you really need a virtual base class?  Two of them??  They almost always lead to trouble and I haven't used it myself (maybe many years ago implementing a c++ library).  I had even forgotten they were part of the language...

Is there an alternative way of implementing the same, or does this really provide a benefit?